feat: add support for TypeScript 4.0

alan-agius4 opened the pull request.

alan-agius4 committed:
feat(compiler-cli): add support for TypeScript 4.0

With this change we add support for TypeScript 4.0

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/metadata.ts:
```diff
@@ -28,7 +28,7 @@ export function generateSetClassMetadataCall(
     return null;
   }
-  const id = ts.updateIdentifier(reflection.getAdjacentNameOfClass(clazz));
+  const id = reflection.getAdjacentNameOfClass(clazz);
```
So as I understand it, `identifiers` never have `typeArguments` (i.e. it is always `undefined`)... right?

alan-agius4 replied:
That is correct, at least I cannot find a case were `typeArguments` is defined.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/switch/src/switch.ts:
```diff
@@ -110,10 +110,6 @@ function flipIvySwitchesInVariableStatement(
-    // Copy the identifier with updateIdentifier(). This copies the internal information which
-    // allows TS to write a correct reference to the identifier.
-    newIdentifier = ts.updateIdentifier(newIdentifier);
-
```
`newIdentifier` can now have a `const` declaration above.

alan-agius4 committed:
refactor(compiler-cli): remove usage of `ts.updateIdentifier`

With Typescript 4, `ts.updateIdentifier` is no longer available.
Calling `ts.updateIdentifier` used to return the same node when
`typeArguments` was `undefined` because `node.typeArguments`
was also `undefined`.

Relevant TS code:
```js
function updateIdentifier(node, typeArguments) {
  return node.typeArguments !== typeArguments
      ? updateNode(createIdentifier(ts.idText(node), typeArguments), node)
      : node;
}
```

alan-agius4 committed:
build: update tslint to 6.1.3

This version supports TypeScript 4.0

alan-agius4 commented:
Requesting reviews and G3 presubmit before TS 4 is final, to get ahead of time.
TS 4 final should be released tomorrow.

kyliau commented:
presubmit: 	http://test/OCL:327476254:BASE:327476296:1597862904109:c514d614

petebacondarwin commented on packages/compiler-cli/src/ngtsc/shims/src/factory_generator.ts:
```diff
@@ -221,7 +219,8 @@ function transformFactorySourceFile(
   }
-  file.statements = ts.createNodeArray(transformedStatements);
+
+  file = ts.updateSourceFileNode(file, transformedStatements);
```
Nice!

petebacondarwin replied:
Is `updateSourceFileNode()` a new API?

alan-agius4 replied:
No, it’s been there for a while.

kyliau replied:
Wow I didn't know this API exists! Always thought source file is immutable and a clone is necessary when modifying a source.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/switch/src/switch.ts:
```diff
@@ -105,7 +104,7 @@ function flipIvySwitchesInVariableStatement(
     // Find the post-switch variable identifier. If one can't be found, it's an error. This is
     // reported as a thrown error and not a diagnostic as transformers cannot output diagnostics.
-    let newIdentifier = findPostSwitchIdentifier(statements, postSwitchName);
+    const newIdentifier = findPostSwitchIdentifier(statements, postSwitchName);
```
How can this be `const`?
The `newIdentifier` variable gets written to below on line 115.

JoostK replied:
@petebacondarwin that code was deleted :D

petebacondarwin replied:
LOL! That was my idea!

petebacondarwin commented on packages/core/schematics/utils/tslint/tslint_html_source_file.ts:
```diff
@@ -17,7 +17,7 @@ export function createHtmlSourceFile(filePath: string, content: string): ts.Sour
   // Subtract two characters because the string literal quotes are only needed for parsing
   // and are not part of the actual source file.
-  sourceFile.end = sourceFile.end - 2;
+  (sourceFile.end as number) = sourceFile.end - 2;
```
OOC - is this because `.end` is not writeable?

alan-agius4 replied:
Yup

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/type_to_value.ts:
```diff
@@ -60,7 +60,7 @@ export function typeToValue(
         // Copying the name here ensures the generated references will be correctly transformed
         // along with the import.
-        expression: ts.updateIdentifier(firstDecl.name),
+        expression: firstDecl.name,
```
The comment above has now become a bit outdated ;)

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/declaration.ts:
```diff
@@ -250,21 +250,10 @@ export class ReturnTypeTransform implements DtsTransform {
-    // `ts.ClassElement`. Since `element` was a `ts.MethodSignature` already, transforming it into
-    // this type is actually correct.
-    return methodSignature as unknown as ts.ClassElement;
+    return ts.updateMethod(
```
Do we know if this still works in TS 3.9? This is a .d.ts transform which was a bit wonky in terms of typings incompatibilities, but this change generates a method declaration an not a method signature as would be the case in the .d.ts, so we should definitely make sure that this still works.

alan-agius4 replied:
We should support 3.9, I’ll verify this tomorrow!

alan-agius4 replied:
tested and works

alan-agius4 commented on packages/compiler-cli/src/metadata/evaluator.ts:
```diff
@@ -476,13 +476,16 @@ export class Evaluator {
-                                       n.kind != ts.SyntaxKind.UndefinedKeyword)
-                               .map(n => this.evaluateNode(n));
+        // TODO(alan-agius4): remove `n.kind !== ts.SyntaxKind.NullKeyword` when
+        // TS 3.9 support is dropped. In TS 4.0 NullKeyword is a child of LiteralType.
```
Will do a follow up

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/declaration.ts:
```diff
@@ -239,32 +239,49 @@ export class ReturnTypeTransform implements DtsTransform {
+    // // TODO(alan-agius4): Remove when we no longer support TS 3.9
+    // TS < 3.9
+    if (ts.isMethodSignature(element)) {
+      const original = ts.getOriginalNode(element, ts.isMethodDeclaration);
```
I suspect this is currently broken. This has to be:

```suggestion
      const original = ts.getOriginalNode(element, ts.isMethodSignature);
```

but I'm not sure if the resulting type of `ts.MethodSignature` would work, given that we previously casted it to `ts.MethodDeclaration` (and now I'm not even sure if that cast was right).

alan-agius4 commented:
@kyliau I made the changes to be compatible with a TS 3.9 compilation, can you please try to run a presubmit?

gkalpak commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -492,11 +492,11 @@ describe('ng program', () => {
       switch (checks.shouldBe) {
         case ShouldBe.Empty:
-          expect(writeData!.data).toMatch(/^(\s*\/\*([^*]|\*[^/])*\*\/\s*)?$/);
+          expect(writeData!.data).toMatch(/^(\s*\/\*([^*]|\*[^\/])*\*\/\s*)?$/);
```
OOC, why this change? It shouldn't be necessary to escape `/` inside `[...]` 😕

alan-agius4 replied:
Pretty sure it was failing without the escaping.

gkalpak replied:
I find that hard to believe 😛
(A little escaping didn't hurt anyone (except for readability maybe 😛), so I don't mind either way :smiley:

gkalpak reviewed (approved):
Commit message typo in **refactor(compiler-cli): remove usage of \`ts.updateIdentifier\`**:
- `unction updateIdentifier` --> `function updateIdentifier`

mhevery reviewed (approved):
reviewed-for: zone-js
reviewed-for: integration-tests

kyliau commented:
presubmit: 	http://test/OCL:327655792:BASE:327655821:1597945695710:c17113a5

alan-agius4 committed:
fixup! feat(compiler-cli): add support for TypeScript 4.0

kyliau commented:
presubmit: 	http://test/OCL:327713907:BASE:327713958:1597964112307:b52bee83

kyliau commented:
g3 presubmit is green!!

crisbeto reviewed (approved):
LGTM

atscott reviewed (changes requested):
reviewed-for: fw-router

alan-agius4 committed:
fixup! feat(compiler-cli): add support for TypeScript 4.0

alan-agius4 committed:
fixup! feat(compiler-cli): add support for TypeScript 4.0

jelbourn reviewed (approved):
LGTM

petebacondarwin reviewed (approved):
Reviewed-for: public-api

alan-agius4 committed:
fixup! feat(compiler-cli): add support for TypeScript 4.0

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

alan-agius4 commented:
Caretaker: angular robot is reporting pending pullapprove status even though everyone approved.

KapilSachdev commented:
It seems 0fc44e0 added `TS 4.0` support to `10.1` and i'm assuming it works fine, but i'm getting the below issue while `ng build --prod`

```js
ERROR in src/app/some/some.component.ts:26:81 - error TS2729: Property 'some_id' is used before its initialization.

26   public data: any = {some_id: this?.some_id, date: new Date()};
                                                                                   ~~~~~~~~~~~~~~~

  src/app/some/some.component.ts:22:12
    22   @Input() some_id
                  ~~~~~~~~~~~~~~~
    'some_id' is declared here.
```

As it is evident, the variable is defined before its being used.
Also it works fine with `3.9`.
_I'm not sure if `TS 4.0` introduced something that could cause this issue, but i think its `ng` so just wanted to confirm. I tried to look into issue tracker for similar reports, though not that deeply, but didn't found any._

fxck commented:
> Property 'serviceStackId$' is used before its initialization.ts(2729)

![image](https://user-images.githubusercontent.com/1303561/95637491-1e9c6100-0a92-11eb-945f-ce7299f9eb7e.png)

happens only with TS 4
